### PR TITLE
fix: corrected uploading for animated images

### PR DESCRIPTION
**Pull request by Gelani-G**
https://jira.railsc.ru/browse/PC4-27093

### CODE REVIEW

File: spec/lib/paperclip/watermark_spec.rb [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  #С путем ВН nil и с типами без ВН
+  let(:instance_with_watermark_path_nil_with_type_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: nil , types_without_watermark: 'image/gif'})
+  end
+  #С путем ВН nil, но без типов
+  let(:instance_with_watermark_path_nil_without_type_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: nil })
+  end
+  #Без пути ВН и без типов
+  let(:instance_without_watermark_path_jpeg) { described_class.new(image_file_jpeg)  }
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:56]: Layout/ExtraSpacing: Unnecessary spacing detected.

File: spec/lib/paperclip/watermark_spec.rb [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  end
+  #С путем ВН nil и с типами без ВН
+  let(:instance_with_watermark_path_nil_with_type_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: nil , types_without_watermark: 'image/gif'})
+  end
+  #С путем ВН nil, но без типов
+  let(:instance_with_watermark_path_nil_without_type_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: nil })
+  end
+  #Без пути ВН и без типов
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:55]: Layout/LeadingCommentSpace: Missing space after #.

File: spec/lib/paperclip/watermark_spec.rb [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  let(:instance_without_watermark_path_with_type_jpeg) do
+    described_class.new(image_file_jpeg, { types_without_watermark: 'image/gif' })
+  end
+  #С путем ВН nil и с типами без ВН
+  let(:instance_with_watermark_path_nil_with_type_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: nil , types_without_watermark: 'image/gif'})
+  end
+  #С путем ВН nil, но без типов
+  let(:instance_with_watermark_path_nil_without_type_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: nil })
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:53]: Style/BracesAroundHashParameters: Redundant curly braces around a hash parameter.<br>Layout/SpaceInsideHashLiteralBraces: Space inside { detected.<br>Layout/SpaceInsideHashLiteralBraces: Space inside } detected.

File: spec/lib/paperclip/watermark_spec.rb [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+  end
+  #С типами без ВН, но без пути ВН
+  let(:instance_without_watermark_path_with_type_jpeg) do
+    described_class.new(image_file_jpeg, { types_without_watermark: 'image/gif' })
+  end
+  #С путем ВН nil и с типами без ВН
+  let(:instance_with_watermark_path_nil_with_type_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: nil , types_without_watermark: 'image/gif'})
+  end
+  #С путем ВН nil, но без типов
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:51]: Layout/LeadingCommentSpace: Missing space after #.

File: spec/lib/paperclip/watermark_spec.rb [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  let(:instance_without_types_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: watermark_path })
+  end
+  #С типами без ВН, но без пути ВН
+  let(:instance_without_watermark_path_with_type_jpeg) do
+    described_class.new(image_file_jpeg, { types_without_watermark: 'image/gif' })
+  end
+  #С путем ВН nil и с типами без ВН
+  let(:instance_with_watermark_path_nil_with_type_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: nil , types_without_watermark: 'image/gif'})
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:49]: Style/BracesAroundHashParameters: Redundant curly braces around a hash parameter.<br>Layout/SpaceInsideHashLiteralBraces: Space inside { detected.<br>Layout/SpaceBeforeComma: Space found before comma.

File: spec/lib/paperclip/watermark_spec.rb [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+  end
+  #С путем ВН без типов
+  let(:instance_without_types_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: watermark_path })
+  end
+  #С типами без ВН, но без пути ВН
+  let(:instance_without_watermark_path_with_type_jpeg) do
+    described_class.new(image_file_jpeg, { types_without_watermark: 'image/gif' })
+  end
+  #С путем ВН nil и с типами без ВН
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:47]: Layout/LeadingCommentSpace: Missing space after #.

File: spec/lib/paperclip/watermark_spec.rb [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+  let(:instance_with_types_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: watermark_path, types_without_watermark: 'image/gif' })
+  end
+  #С путем ВН без типов
+  let(:instance_without_types_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: watermark_path })
+  end
+  #С типами без ВН, но без пути ВН
+  let(:instance_without_watermark_path_with_type_jpeg) do
+    described_class.new(image_file_jpeg, { types_without_watermark: 'image/gif' })
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:45]: Style/BracesAroundHashParameters: Redundant curly braces around a hash parameter.<br>Layout/SpaceInsideHashLiteralBraces: Space inside { detected.<br>Layout/SpaceInsideHashLiteralBraces: Space inside } detected.

File: spec/lib/paperclip/watermark_spec.rb [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  #ИНСТАНСЫ С ФАЙЛОМ РАСШИРЕНИЯ JPEG (ВН - ВОДЯНОЙ ЗНАК)
+  #С путем ВН и типами без ВН
+  let(:instance_with_types_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: watermark_path, types_without_watermark: 'image/gif' })
+  end
+  #С путем ВН без типов
+  let(:instance_without_types_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: watermark_path })
+  end
+  #С типами без ВН, но без пути ВН
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:43]: Layout/LeadingCommentSpace: Missing space after #.

File: spec/lib/paperclip/watermark_spec.rb [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+  #Без пути ВН и без типов
+  let(:instance_without_watermark_path_gif) { described_class.new(image_file_gif) }
+  #ИНСТАНСЫ С ФАЙЛОМ РАСШИРЕНИЯ JPEG (ВН - ВОДЯНОЙ ЗНАК)
+  #С путем ВН и типами без ВН
+  let(:instance_with_types_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: watermark_path, types_without_watermark: 'image/gif' })
+  end
+  #С путем ВН без типов
+  let(:instance_without_types_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: watermark_path })
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:41]: Style/BracesAroundHashParameters: Redundant curly braces around a hash parameter.<br>Layout/SpaceInsideHashLiteralBraces: Space inside { detected.<br>Layout/SpaceInsideHashLiteralBraces: Space inside } detected.

File: spec/lib/paperclip/watermark_spec.rb [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    described_class.new(image_file_gif, { watermark_path: nil })
+  end
+  #Без пути ВН и без типов
+  let(:instance_without_watermark_path_gif) { described_class.new(image_file_gif) }
+  #ИНСТАНСЫ С ФАЙЛОМ РАСШИРЕНИЯ JPEG (ВН - ВОДЯНОЙ ЗНАК)
+  #С путем ВН и типами без ВН
+  let(:instance_with_types_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: watermark_path, types_without_watermark: 'image/gif' })
+  end
+  #С путем ВН без типов
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:39]: Layout/LeadingCommentSpace: Missing space after #.

File: spec/lib/paperclip/watermark_spec.rb [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+RSpec.describe Paperclip::Watermark do
+  let(:image_file_jpeg) do
+    fixture_file_upload(Rails.root.join('../fixtures/images/sample_image.jpg'), 'image/jpeg', :binary)
+  end
+  let(:image_file_gif) do
+    fixture_file_upload(Rails.root.join('../fixtures/images/sample_gif_image.gif'), 'image/gif', :binary)
+  end
+  let(:watermark_path) { Rails.root.join('../fixtures/images/pcwm_small.png') }
+  #ИНСТАНСЫ С ФАЙЛОМ РАСШИРЕНИЯ GIF (ВН - ВОДЯНОЙ ЗНАК)
```
Comment (taleksei) [inline spec/lib/paperclip/watermark_spec.rb:11]: а почему не ВЗ?

думаю не надо поясняющих коментоа, т.к. понятно по названию, иначе начем такие подробные завания)

File: spec/lib/paperclip/watermark_spec.rb [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    described_class.new(image_file_jpeg, { watermark_path: watermark_path, types_without_watermark: 'image/gif' })
+  end
+  let(:instance_without_types_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: watermark_path })
+  end
+  let(:instance_without_watermark_path_with_type_jpeg) do
+    described_class.new(image_file_jpeg, { types_without_watermark: 'image/gif' })
+  end
+  let(:instance_with_watermark_path_nil_with_type_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: nil, types_without_watermark: 'image/gif' })
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:37]: Style/BracesAroundHashParameters: Redundant curly braces around a hash parameter.<br>Layout/SpaceInsideHashLiteralBraces: Space inside { detected.<br>Layout/SpaceInsideHashLiteralBraces: Space inside } detected.

File: spec/lib/paperclip/watermark_spec.rb [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+  end
+  let(:instance_with_watermark_path_nil_with_type_gif) do
+    described_class.new(image_file_gif, { watermark_path: nil, types_without_watermark: 'image/gif' })
+  end
+  let(:instance_with_watermark_path_nil_without_type_gif) do
+    described_class.new(image_file_gif, { watermark_path: nil })
+  end
+  let(:instance_without_watermark_path_gif) { described_class.new(image_file_gif) }
+  let(:instance_with_types_jpeg) do
+    described_class.new(image_file_jpeg, { watermark_path: watermark_path, types_without_watermark: 'image/gif' })
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:28]: Style/BracesAroundHashParameters: Redundant curly braces around a hash parameter.<br>Layout/SpaceInsideHashLiteralBraces: Space inside { detected.<br>Layout/SpaceInsideHashLiteralBraces: Space inside } detected.

File: spec/lib/paperclip/watermark_spec.rb [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+    described_class.new(image_file_gif, { watermark_path: watermark_path })
+  end
+  let(:instance_without_watermark_path_with_type_gif) do
+    described_class.new(image_file_gif, { types_without_watermark: 'image/gif' })
+  end
+  let(:instance_with_watermark_path_nil_with_type_gif) do
+    described_class.new(image_file_gif, { watermark_path: nil, types_without_watermark: 'image/gif' })
+  end
+  let(:instance_with_watermark_path_nil_without_type_gif) do
+    described_class.new(image_file_gif, { watermark_path: nil })
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:24]: Style/BracesAroundHashParameters: Redundant curly braces around a hash parameter.<br>Layout/SpaceInsideHashLiteralBraces: Space inside { detected.<br>Layout/SpaceInsideHashLiteralBraces: Space inside } detected.

File: spec/lib/paperclip/watermark_spec.rb [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    described_class.new(image_file_gif, { watermark_path: watermark_path, types_without_watermark: 'image/gif' })
+  end
+  let(:instance_without_types_gif) do
+    described_class.new(image_file_gif, { watermark_path: watermark_path })
+  end
+  let(:instance_without_watermark_path_with_type_gif) do
+    described_class.new(image_file_gif, { types_without_watermark: 'image/gif' })
+  end
+  let(:instance_with_watermark_path_nil_with_type_gif) do
+    described_class.new(image_file_gif, { watermark_path: nil, types_without_watermark: 'image/gif' })
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:21]: Style/BracesAroundHashParameters: Redundant curly braces around a hash parameter.<br>Layout/SpaceInsideHashLiteralBraces: Space inside { detected.<br>Layout/SpaceInsideHashLiteralBraces: Space inside } detected.

File: spec/lib/paperclip/watermark_spec.rb [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+  end
+  let(:watermark_path) { Rails.root.join('../fixtures/images/pcwm_small.png') }
+  let(:instance_with_types_gif) do
+    described_class.new(image_file_gif, { watermark_path: watermark_path, types_without_watermark: 'image/gif' })
+  end
+  let(:instance_without_types_gif) do
+    described_class.new(image_file_gif, { watermark_path: watermark_path })
+  end
+  let(:instance_without_watermark_path_with_type_gif) do
+    described_class.new(image_file_gif, { types_without_watermark: 'image/gif' })
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:18]: Style/BracesAroundHashParameters: Redundant curly braces around a hash parameter.<br>Layout/SpaceInsideHashLiteralBraces: Space inside { detected.<br>Layout/SpaceInsideHashLiteralBraces: Space inside } detected.

File: spec/lib/paperclip/watermark_spec.rb [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+  end
+  let(:image_file_gif) do
+    fixture_file_upload(Rails.root.join('../fixtures/images/sample_gif_image.gif'), 'image/gif', :binary)
+  end
+  let(:watermark_path) { Rails.root.join('../fixtures/images/pcwm_small.png') }
+  let(:instance_with_types_gif) do
+    described_class.new(image_file_gif, { watermark_path: watermark_path, types_without_watermark: 'image/gif' })
+  end
+  let(:instance_without_types_gif) do
+    described_class.new(image_file_gif, { watermark_path: watermark_path })
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:15]: Style/BracesAroundHashParameters: Redundant curly braces around a hash parameter.<br>Layout/SpaceInsideHashLiteralBraces: Space inside { detected.<br>Layout/SpaceInsideHashLiteralBraces: Space inside } detected.

File: spec/lib/paperclip/watermark_spec.rb [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+RSpec.describe Paperclip::Watermark do
+  let(:image_file_jpeg) do
+    fixture_file_upload(Rails.root.join('../fixtures/images/sample_image.jpg'), 'image/jpeg', :binary)
+  end
+  let(:image_file_gif) do
+    fixture_file_upload(Rails.root.join('../fixtures/images/sample_gif_image.gif'), 'image/gif', :binary)
+  end
+  let(:watermark_path) { Rails.root.join('../fixtures/images/pcwm_small.png') }
+  let(:instance_with_types_gif) do
+    described_class.new(image_file_gif, { watermark_path: watermark_path, types_without_watermark: 'image/gif' })
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:12]: Style/BracesAroundHashParameters: Redundant curly braces around a hash parameter.<br>Layout/SpaceInsideHashLiteralBraces: Space inside { detected.<br>Layout/SpaceInsideHashLiteralBraces: Space inside } detected.

File: lib/paperclip/watermark.rb [hunk 19]
```diff
@@ -34,8 +35,10 @@ def make
         Paperclip.run(program, parameters,
           :source => "#{File.expand_path(src.path)}#{'[0]' unless animated?}",
           :dest => File.expand_path(dst.path),
-          :watermark => watermark_path
-        )
+          :watermark => !file.is_a?(Paperclip::Tempfile) ?
+            (types_without_watermark.exclude?(file.content_type) ? watermark_path : nil) :
+            watermark_path
+            )
```
Comment (apress-hound) [inline lib/paperclip/watermark.rb:41]: Layout/ClosingParenthesisIndentation: Indent ) to column 8 (not 12)<br>Layout/MultilineMethodCallBraceLayout: Closing method call brace must be on the same line as the last argument when opening brace is on the same line as the first argument.

File: lib/paperclip/watermark.rb [hunk 20]
```diff
@@ -34,8 +35,10 @@ def make
         Paperclip.run(program, parameters,
           :source => "#{File.expand_path(src.path)}#{'[0]' unless animated?}",
           :dest => File.expand_path(dst.path),
-          :watermark => watermark_path
-        )
+          :watermark => !file.is_a?(Paperclip::Tempfile) ?
+            (types_without_watermark.exclude?(file.content_type) ? watermark_path : nil) :
```
Comment (apress-hound) [inline lib/paperclip/watermark.rb:39]: Style/NestedTernaryOperator: Ternary operators must not be nested. Prefer if or else constructs instead.

File: lib/paperclip/watermark.rb [hunk 21]
```diff
@@ -34,8 +35,10 @@ def make
         Paperclip.run(program, parameters,
           :source => "#{File.expand_path(src.path)}#{'[0]' unless animated?}",
           :dest => File.expand_path(dst.path),
-          :watermark => watermark_path
-        )
+          :watermark => !file.is_a?(Paperclip::Tempfile) ?
```
Comment (apress-hound) [inline lib/paperclip/watermark.rb:38]: Style/MultilineTernaryOperator: Avoid multi-line ternary operators, use if or unless instead.

File: spec/lib/paperclip/watermark_spec.rb [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+    described_class.new(image_file_jpeg, watermark_path: watermark_path)
+  end
+  let(:instance_without_watermark_path_with_type_jpeg) do
+    described_class.new(image_file_jpeg, types_without_watermark: 'image/gif')
+  end
+  let(:instance_with_watermark_path_nil_with_type_jpeg) do
+    described_class.new(image_file_jpeg, watermark_path: nil, types_without_watermark: 'image/gif')
+  end
+  let(:instance_with_watermark_path_nil_without_type_jpeg) do
+    described_class.new(image_file_jpeg, watermark_path: nil )
```
Comment (apress-hound) [inline spec/lib/paperclip/watermark_spec.rb:40]: Layout/SpaceInsideParens: Space inside parentheses detected.

**Comment by TamarinEA** [comment]
прикрепи задачу пожалуйста

**Comment by TamarinEA** [comment]
задачу кладём в первый комментарий, а не название пр

обычно, если коммит назвал правильно и заслал только его (сейчас у тебя правильный коммит) - то и название и первый комментарий автоматически заполняются

**Review by taleksei — APPROVED** [review]

**Comment by taleksei** [comment]
@c1n1c

**Comment by c1n1c** [comment]
и если задача в сентябрь-1 то сразу пиши, сделайте ветку) а то на автомате примем не заметим, а ветка должна быть pc-september-1

**Review by c1n1c — APPROVED** [review]

**Pull request merged**